### PR TITLE
Serve correctly hashed static files with the Cache-Control header

### PR DESCRIPTION
**Pull request by eashwar**
Issue: #8278 

By @eashwar and @christopherhafke.

## What we did

After reproducing the behavior described in #8278, we added the `Cache-Control` header to both the static files in the user-specified `staticDir` as well as any files served from `/static/media` (i.e., files loaded and hashed by `file-loader`).

## How to test

To test this change, one can write any story that uses some static file (like an image) located in the `public` or `src` directories of a project. An example of a story that we wrote for `examples/cra-react15`:

```
import logo from '../logo.svg';
...
export const Story3 = () => (
  <Button onClick={action('clicked')}>
    <img src={logo} alt='A nice logo.'/>
  </Button>
);
Story3.story = { name: 'with an image'};
```
Run `yarn storybook`, go to the story (with browser devtools open). The response header for the image file should have the header `Cache-Control`. Reload the page, and the network request made for the image (in this example, `logo.svg`), should say something along the lines of `200 (from memory/disk cache)` in the response header (as opposed to `304 not modified`).

### Is this testable with Jest or Chromatic screenshots?

We're unsure about how we could test something like this with Jest, or if that needs to be done at all - `build-dev.js` and `dev-server.js` don't seem to have tests written for them.

### Does this need a new example in the kitchen sink apps?

This is unlikely, as the PR does not add any new functionality (instead improves existing functionality).

### Does this need an update to the documentation?

No.

### CODE REVIEW

**Comment by eashwar** [comment]
It looks like the build is failing - we rebased onto the (as of writing this message) latest commit on `next`, which also seems to be failing. Should we/will we have to rebase again when `next` reaches a non-failing state?

**Comment by ndelangen** [comment]
So when a user DOES make a change to the file, will it still show the new version?

**Comment by christopherhafke** [comment]
@ndelangen Yes, when changes are made (e.g. edits, creation, deletion) the new changes are correctly pulled in with the expected `200 OK` response (rather than the `200 OK (from disk cache)` or `200 OK (from memory cache)` response from a cached asset).

**Comment by eashwar** [comment]
However, this doesn't work for files _without_ the hash in their name (i.e., files in the `staticDir` directories specified by the user like the `public` folder in a CRA app). Notably, adding this change to things that aren't in the `/static/media` folder wasn't actually asked for in the issue #8278 so we're going to remove that addition (but keep the added header for hashed files).

File: lib/core/src/server/dev-server.js [hunk 1]
```diff
@@ -61,6 +61,11 @@ export default function (options) {
       devMiddlewareOptions
     );
 
+    router.get(/\/static\/media\/.*\..*/, (request, response, next) => {
+      response.set('Cache-Control', `public, max-age=31536000`);
```
Comment (ndelangen) [inline lib/core/src/server/dev-server.js:65]: Ow I see, if there's a hash in the filename, then the hash changes when the file changes, so we can add a long term cache.

Comment (ndelangen) [reply lib/core/src/server/dev-server.js:65]: How is this not applied to files without a hash?

Comment (eashwar) [reply lib/core/src/server/dev-server.js:65]: Essentially, we add this header only to resources under the `/static/media/` path; users don't really have the ability to directly put files there without them being put there by `file-loader` or `media-loader`. As a result, anything served from `/static/media` _will_ be a file that's hashed, so we can apply this header to any file served from here.

`file-loader` and `media-loader` are configured to use this path in [`manager-webpack.config.js:115`](https://github.com/storybookjs/storybook/blob/34ec239aa2b0ceb022d18e6c3e45ab8ca797faa7/lib/core/src/server/manager/manager-webpack.config.js#L115)

**Review by ndelangen — APPROVED** [review]

**Comment by donaldpipowitch** [comment]
Is there a way to get this behaviour also for files covered by `staticDir`? I have fonts coming from there as well. I also tried to putting them into a `/static/media` directory, but files coming from `/:staticDir/static/media` don't seem to get the cache control header. It's really just for files which were directly referenced by `file-loader` if I'm not mistaken.

**Comment by shilman** [comment]
cc @eashwar @ndelangen

**Pull request merged**